File: algorithms/sonar-visit.js
/**
 * Sonar Visit algorithm - O(1) angle sweep approach
 * Compatible with both browser and Node.js
 * @see https://github.com/konard/sonar
 */

(function (root, factory) {
  if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  } else {
    root.SonarVisit = factory();
  }
}(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  /**
   * Generate a tour using the Sonar Visit algorithm
   * 360 degree rotation sweep in fixed angle steps
   * Points on the same angle are connected by distance from center
   * Time complexity: O(1) angle steps (determined by grid size, not point count)
   *
   * @param {Array} points - Array of point objects with x, y, angle, and id properties
   * @param {number} gridSize - Grid size used for angle bucket calculation (default: 40)
   * @returns {Array<number>} Tour as array of point indices
   */
  function generateTour(points, gridSize = 40) {
    // Calculate angle step based on grid size
    const angleSteps = 4 * gridSize;
    const angleStep = (2 * Math.PI) / angleSteps;

    // Group points by their angle bucket
    const angleBuckets = new Map();

    points.forEach(point => {
      let angle = point.angle;
      if (angle < 0) angle += 2 * Math.PI;

      const bucketIndex = Math.floor(angle / angleStep);
      if (!angleBuckets.has(bucketIndex)) {
        angleBuckets.set(bucketIndex, []);
      }
      angleBuckets.get(bucketIndex).push(point);
    });

    // Sort points within each bucket by distance from center
    const center = { x: 0.5, y: 0.5 };
    angleBuckets.forEach((bucketPoints, key) => {
      bucketPoints.sort((a, b) => {
        const distA = Math.hypot(a.x - center.x, a.y - center.y);
        const distB = Math.hypot(b.x - center.x, b.y - center.y);
        return distA - distB;
      });
    });

    // Build tour by sweeping through angle buckets in order
    const tour = [];
    for (let i = 0; i < angleSteps; i++) {
      if (angleBuckets.has(i)) {
        const bucketPoints = angleBuckets.get(i);
        bucketPoints.forEach(p => tour.push(p.id));
      }
    }

    return tour;
  }

  return {
    generateTour
  };
}));

File: rust/src/sonar_visit.rs
//! Sonar Visit algorithm - O(1) angle sweep approach

use std::collections::BTreeMap;
use crate::utils::Point;
use std::f64::consts::PI;

/// Generate a tour using the Sonar Visit algorithm
/// 360 degree rotation sweep in fixed angle steps
/// Points on the same angle are connected by distance from center
/// Time complexity: O(1) angle steps (determined by grid size, not point count)
pub fn generate_tour(points: &[Point], grid_size: usize) -> Vec<usize> {
    // Calculate angle step based on grid size
    let angle_steps = 4 * grid_size;
    let angle_step = (2.0 * PI) / angle_steps as f64;

    // Group points by their angle bucket
    let mut angle_buckets: BTreeMap<usize, Vec<&Point>> = BTreeMap::new();

    for point in points {
        let mut angle = point.angle;
        if angle < 0.0 {
            angle += 2.0 * PI;
        }

        let bucket_index = (angle / angle_step) as usize;
        angle_buckets.entry(bucket_index).or_default().push(point);
    }

    // Sort points within each bucket by distance from center
    let center_x = 0.5;
    let center_y = 0.5;

    for bucket_points in angle_buckets.values_mut() {
        bucket_points.sort_by(|a, b| {
            let dist_a = ((a.x - center_x).powi(2) + (a.y - center_y).powi(2)).sqrt();
            let dist_b = ((b.x - center_x).powi(2) + (b.y - center_y).powi(2)).sqrt();
            dist_a.partial_cmp(&dist_b).unwrap()
        });
    }

    // Build tour by sweeping through angle buckets in order
    let mut tour = Vec::with_capacity(points.len());
    for i in 0..angle_steps {
        if let Some(bucket_points) = angle_buckets.get(&i) {
            for p in bucket_points {
                tour.push(p.id);
            }
        }
    }

    tour
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::generate_normalized_points;

    #[test]
    fn test_sonar_visit_visits_all() {
        let points = generate_normalized_points(10, 20, 42);
        let tour = generate_tour(&points, 20);

        assert_eq!(tour.len(), points.len());

        // Check no duplicates
        let mut seen = std::collections::HashSet::new();
        for &id in &tour {
            assert!(seen.insert(id));
        }
    }
}
